sqlx-core/src/sqlite/connection/collation.rs

```diff
@@ -0,0 +1,82 @@
+use std::cmp::Ordering;
+use std::ffi::CString;
+use std::os::raw::{c_char, c_int, c_void};
+use std::panic::{catch_unwind, UnwindSafe};
+use std::slice;
+
+use libsqlite3_sys::{sqlite3_create_collation_v2, SQLITE_OK, SQLITE_UTF8};
+
+use crate::error::Error;
+use crate::sqlite::connection::handle::ConnectionHandle;
+use crate::sqlite::SqliteError;
+
+unsafe extern "C" fn free_boxed_value<T>(p: *mut c_void) {
+    drop(Box::from_raw(p as *mut T));
+}
+
+pub(crate) fn create_collation<F>(
+    handle: &ConnectionHandle,
+    name: &str,
+    collation: F,
+) -> Result<(), Error>
+where
+    F: Fn(&str, &str) -> Ordering + Send + Sync + UnwindSafe + 'static,
+{
+    unsafe extern "C" fn call_boxed_closure<C>(
+        arg1: *mut c_void,
+        arg2: c_int,
+        arg3: *const c_void,
+        arg4: c_int,
+        arg5: *const c_void,
+    ) -> c_int
+    where
+        C: Fn(&str, &str) -> Ordering,
+    {
+        let r = catch_unwind(|| {
+            let boxed_f: *mut C = arg1 as *mut C;
+            assert!(!boxed_f.is_null(), "Internal error - null function pointer");
```

Comment:
I would remove the message here and downgrade this to `debug_assert!`. This is a programming error of _SQLx_. The user can't trigger this.